### DISCUSSION TITLE
Redundancy

### DISCUSSION THREAD

**Discussion by robsori** [frangoteam/FUXA#809] [Ideas]
In any SCADA application, the redundancy scenario is required. This implies the following:
1. Two servers in operation, one active with main role and one hot reserve with secondary role
2. Fuxa clients connected to a single Fuxa server, the active one.
When the active Fuxa server fails, clients must automatically switch to the hot reserve Fuxa server so that the operators supervising a technological process are not affected and remain without synoptic screens.
After repairing the active Fuxa server and connecting to the data network, the databases of the two servers must be synchronized. Fuxa clients then switch to the returned active server.
Can we implement such a scenario on Fuxa? Do you have other ideas?

**Comment by rvbatista** [comment]
For the communication it's not that simple, some protocols, physical layer or devices isn't able to communicate at the same time.
Redundancy requires a high level of decoupling to allow communication, alarms processing, trending, etc.. processes communicate with each other through Network.

But an initial approach for redundancy is client code for server switching and server state synchronization, it will create some communication problems but only for some communication protocols.

**Comment by unocelli** [comment]
The feature is interesting but I agree with @rvbatista a complete redundancy is very challenging.

**Reply by robsori** [reply]
I'm talking about the redundancy scenario with two active Fuxa servers (mirror image).
The permanent communication of the two independent Fuxa servers, active and hot reserve, do you think they would face simultaneous communication problems with field equipment (PLCs for example)? Each Fuxa server will be installed on a different computer. I think not.
I think that the first step to solve would be the redirection of the client side to the functional backup server (hot backup) when the first one fails. Perhaps, in this scenario, a third-party application should be installed on the computer on which the Fuxa client is called to automatically switch the opening of the browser to the other server. The operator will see the opening of a new session to the server for a few seconds.
The third application will have to constantly monitor the two Fuxa servers and make this switch.
The next step would be to resolve the database synchronization between the two Fuxa servers.
It's just an idea.

**Reply by rvbatista** [reply]
Today FUXA have Modbus RTU, over serial line and over TCP/IP, both needs some kind of exclusive usage over the connection to avoid collisions.
Also some simple TCP devices can have limited amount of connections that they can handle, with "unknown" consequences with you go over that connection limit.
So it's good to have a kind of logic that only one server is allowed to connect to the device.
For trend data sync, a good solution is to both servers store data on the same server, but only the one active.

And yes, it's a good feature, but I think that it's a lot of work to make it work perfectly.
